Checkout: Add password strength meter to new field

mikejolley opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

Adds a password strength meter to the password field on checkout:

![Screenshot 2024-07-04 at 17 31 49](https://github.com/woocommerce/woocommerce/assets/90977/91762efe-5f43-42b4-b200-193a120c761a)

With validation error:

![Screenshot 2024-07-05 at 11 32 52](https://github.com/woocommerce/woocommerce/assets/90977/c7a0a246-8c65-439f-a047-1712a2daee8b)

Key details about the implementation:

- Introduced the ability to pass in the feedback component to `ValidatedTextField` so we can inject the strength meter
- Enqueues `zxcvbn-async` if registration is enabled and the user is logged out. If `zxcvbn` is not present, it will instead use the `check-password-strength` library which is less accurate. We could instead just hide the meter if zxcvbn is not present (feedback?)
- Uses the semantic HTML `meter` component. Browser compatibility was awkward but I tested across Safari, Arc, Firefox. [Firefox does not show the "score" because `:after` is not supported.](https://css-tricks.com/html5-meter-element/)
- I've set a minimum allowed strength of 2 (medium) which will block checkout if not met. Is this sensible default or do we want it lower or unrestricted?

Closes #48628

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. In store Account and Privacy settings, enable registration during checkout and turn off the option to automatically generate passwords
2. As a guest customer, add something to cart and go to checkout.
3. Check the "create account" checkbox
4. Enter a password. Test a variety of passwords and ensure the strength indicator changes 
5. Try to place order with weak password. You should see a validation error.
6. Place order with strong password. Ensure account was created.

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

mikejolley committed:
Add validation message and fix rendering when account is required

mikejolley committed:
Adjust validation so we can change the label in messages with custom callback

mikejolley committed:
Add and style password meter

mikejolley committed:
Add validation feedback and zxcvbn support

mikejolley committed:
changelog

mikejolley commented:
I don't think the test fails are related to these changes so I'll mark for review. Note that this targets the https://github.com/woocommerce/woocommerce/pull/48985 branch, so I'll merge this into the other branch before trunk, or rebase on trunk if we do the other one first.

senadir commented:
Some notes about your questions:

- Enqueues zxcvbn-async if registration is enabled and the user is logged out. If zxcvbn is not present, it will instead use the check-password-strength library which is less accurate. We could instead just hide the meter if zxcvbn is not present (feedback?)
  -  We can look into making this opt-in at some point? I still hold the opinion that this shouldn't be strict for all stores.
- Uses the semantic HTML meter component. Browser compatibility was awkward but I tested across Safari, Arc, Firefox. [Firefox does not show the "score" because :after is not supported.](https://css-tricks.com/html5-meter-element/)
  - Can we detect if `after` is not supported and just do a couple of components. Doesn't seem like a huge problem for now.
- I've set a minimum allowed strength of 2 (medium) which will block checkout if not met. Is this sensible default or do we want it lower or unrestricted?
  - We can start with keeping this set in stone as 2, if we ever make it extensible (filter maybe in which store sets an accepted strength), we can reevaluate, if it's less than 2, we should just skip the strength meter completely, no point from using it if everything is accepted.

mikejolley commented:
> We can look into making this opt-in at some point? I still hold the opinion that this shouldn't be strict for all stores.

Yeah but I still find the protection against common passwords really beneficial to secure accounts. 

> Can we detect if after is not supported and just do a couple of components. Doesn't seem like a huge problem for now.

We could avoid using meter and use something custom, or append the label some other way. I was trying to keep it semantic. The bar should still be understandable without the text.

> We can start with keeping this set in stone as 2, if we ever make it extensible (filter maybe in which store sets an accepted strength), we can reevaluate, if it's less than 2, we should just skip the strength meter completely, no point from using it if everything is accepted.

Makes sense.

opr reviewed (changes requested):
Nice work on this Mike, it's looking good. The code looks alright, but I didn't do a full review on the CSS, because I think things will change following this review.

A few things I spotted and wonder about:

- The layout shifts when the meter appears, could we either:
  - Keep the meter visible, but indicate it'll update as they type (if necessary)
  - Keep it hidden, but reserve the space for the meter, so when it appears there's no shift
- Is this accessible, would a screen reader know about the password strength while typing? While using VoiceOver, the password strength was not read out to me. (actually, voicover says "Two shipping options were found" while interacting with this field, but I doubt it's related to the strength meter PR). Maybe consider having the strength announced to screen readers after typing?
- On mobile/small desktop, if the "Please create a stronger password" text is long (e.g. after being translated) then it overlaps with the strength.
<img width="369" alt="image" src="https://github.com/woocommerce/woocommerce/assets/5656702/d60c766a-8e86-4fe6-8a6b-028cfa1784be">

> The bar should still be understandable without the text.

I'm not confident about this, unless we add some text indicating it's for password strength.

mikejolley committed:
Accessibility fixes

mikejolley committed:
Update lock file

mikejolley commented:
@opr 

- Added feedback element that shows strength across all browsers and is picked up by voiceover
- Prevented error notices and strength bar rendering at the same time
- Updated initial state to be hidden but take up space

opr reviewed (changes requested):
Thanks for the changes!

A11y, think this is working OK but is it possible to read the strength with "Password strength: %s" rather than just the strength? I feel like it could be confusing to just hear "Weak" "medium" with no real context. 

Is it also possible to delay this announcement until the user has finished typing, or at least announce it every time the password changes, not just the first time it switches strength level?

There's still some layout shift when tabbing out of the input with an invalid password

https://www.loom.com/share/f9a9d4113e364a48b5dac7686bde27ad

In Safari, the "weak" bar is green and bigger than it should be
<img width="817" alt="image" src="https://github.com/woocommerce/woocommerce/assets/5656702/9d92d7b9-2dc2-4961-8b6c-d8da542a377d">

mikejolley committed:
Fix initial state in safari

mikejolley committed:
Announce strength as you type

mikejolley commented:
@opr 

> A11y, think this is working OK but is it possible to read the strength with "Password strength: %s" rather than just the strength? I feel like it could be confusing to just hear "Weak" "medium" with no real context.

I've added that and the character count so its announced more often.

> There's still some layout shift when tabbing out of the input with an invalid password

All fields with validation notices have layout shift.

> In Safari, the "weak" bar is green and bigger than it should be

Should be fixed. Safari has some unique bugs around the meter element.

opr commented on plugins/woocommerce-blocks/assets/js/blocks/checkout/password-strength-meter/index.tsx:
```diff
@@ -0,0 +1,119 @@
+}: {
+	password: string;
+	onChange?: ( strength: number ) => void;
+} ): JSX.Element | null => {
```
`JSX.Element` is deprecated and the preferred way is `React.ReactElement` though I don't think this should block this PR, feel free to change it if you like.

opr reviewed (approved):
Thanks Mike, this is working well for me, a11y is improved a lot (though still facing the issue with the shipping options being read out, but that's separate) and safari looks to be working well too.

Happy for you to merge this and resolve the failing tests on the base PR.

mikejolley committed:
Use React.ReactElement

mikejolley committed:
update lock file

mikejolley committed:
Update lock

mikejolley committed:
Downgrade local pnpm and recreate lock

mikejolley committed:
Feedback should only be shown when `showError` is true, not `hasError`

mikejolley committed:
Feedback should default to null
